addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py

```diff
@@ -209,37 +218,60 @@ def _l10n_sa_get_monetary_vals(self, invoice, vals):
         # We use base_amount_currency + tax_amount_currency instead of amount_total because we do not want to include
         # withholding tax amounts in our calculations
         total_amount = abs(vals['taxes_vals']['base_amount_currency'] + vals['taxes_vals']['tax_amount_currency'])
-
+        line_extension_amount = vals['vals']['legal_monetary_total_vals']['line_extension_amount']
         tax_inclusive_amount = total_amount
         tax_exclusive_amount = abs(vals['taxes_vals']['base_amount_currency'])
         prepaid_amount = 0
         payable_amount = total_amount
-
         # - When we calculate the tax values, we filter out taxes and invoice lines linked to downpayments.
         #   As such, when we calculate the TaxInclusiveAmount, it already accounts for the tax amount of the downpayment
         #   Same goes for the TaxExclusiveAmount, and we do not need to add the Tax amount of the downpayment
         # - The payable amount does not account for the tax amount of the downpayment, so we add it
         downpayment_vals = self._l10n_sa_get_prepaid_amount(invoice, vals)
-
+        allowance_charge_vals = vals['vals']['allowance_charge_vals']
+        allowance_total_amount = sum(a['amount'] for a in allowance_charge_vals if a['charge_indicator'] == 'false')
         if downpayment_vals:
-            # Makes no sense, but according to ZATCA, if there is a downpayment, the TotalInclusiveAmount
-            # should include the total amount of the invoice (including downpayment amount) PLUS the downpayment
-            # total amount, AGAIN.
-            prepaid_amount = tax_inclusive_amount + downpayment_vals['total_amount']
-            payable_amount = - downpayment_vals['total_amount']
-
+            prepaid_amount = downpayment_vals['total_amount']
+            payable_amount -= prepaid_amount
         return {
+            'line_extension_amount': line_extension_amount - allowance_total_amount,
```

Comment:
I wonder if that should not be in generic then?

Reply:
It should be, indeed